### DISCUSSION TITLE
Runtime logs not showing

### DISCUSSION THREAD

**Discussion by Martin-Carlsson** [vercel/community#4559] [Bug Report]
### Page to Investigate

https://vercel.com/martin-carlsson/meritmeadow/logs?page=1&timeline=past30Minutes&startDate=1697867805158&endDate=1697869605158

### Steps to Reproduce

Runtime logs not showing from `console.log` are not showing in Vercel Logs dashboard.

The Vercel Logs dashboard are showing other logs like [GET] request, but everything that comes from `console.log` are not showing.

The logs were showing fine yesterday.

**Comment by Martin-Carlsson** [comment]
Now it's back.

**Comment by Martin-Carlsson** [comment]
And now it's gone again.

**Comment by Martin-Carlsson** [comment]
It turned out that the Log sometimes combines all `console.log` messages in to one message.

So, sometimes I get `console.log` messages one by one, and sometimes they are combined in one large message.

The prefered behaviour would be to have messages shown one by one, for two reasons:
 - In a debugging phase you will have **a lot** of `console.log` messages, but in a combined message, the message is abbridge if it goes above a certain length, so I don't get to see all data.
 - The order of messages is important.
 
 **For internal Vercel debugging:**
 
 - **Request id** `mt5lt-1697873914840-9859d12e8126` is an example of a combined message 😭
 - **Request id** `m98hs-1697873489091-4672e1a5a08e` is an example of an individual message 👌🤩🎉

**Comment by Martin-Carlsson** [comment]
After a few too many hours spent on this issue ...

It looks like Vercel deployment has a random element build in. Most of the time you get a deploy where all messages are combined, and some time you are lucky and get a line by line.

I don't love that deploys behaves differently from each other.

For anyone reading this with the same issue I found two workarounds:

1. Keep deploying until you get a good deploy. Just change a comma or something in your frontend and push it to your repo. Redeploy doesn't seem to change anything, but I may be wrong.
2. Install the Vercel CLI and run `vercel logs <DEPLOYMENT_ID> --follow`

**Comment by zvictor** [comment]
For me, the logs never seem to work. I tried setting up Sematext hoping that logs would appear there, but no success

**Comment by vasanth-asokan** [comment]
Related to this thread, in my case a Python runtime -- no logs are showing up when using Python's `logging` but print works. I've `logging` directly emitting to logtail as well, and they don't show up on logtail either but just in Vercel Production. Emitting to logtail works fine locally when running `vercel dev` or the server directly, it's bizarre how it just doesn't work in Vercel production.

It's almost like the Vercel production runtime is taking over python logging in some way, or setting a level to something above INFO!

On a tangential note, Vercel documentation is sparse on what to expect with user generated / custom log messages, how to access them, constraints etc.

**Reply by Martin-Carlsson** [reply]
That's just very strange behaviour 🤔

**Comment by kapitanluffy** [comment]
@Martin-Carlsson so you're saying reroll the deployment and check the log hoping it will show up what you need?

That seems.. brittle..

**Reply by Martin-Carlsson** [reply]
That was my conclusion at the time

That is.. fickle..

**Comment by enesakar** [comment]


**Reply by leerob** [reply]
We don't have any reported issues right now. Could you open a support ticket please with what you are seeing?

**Comment by masonmcelvain** [comment]
Similar, though not the same as, this issue: https://github.com/vercel/next.js/issues/61240

**Comment by saguilarolmo** [comment]
I have encountered a similar problem. I have a serveless function that runs through a Cron Job every minute. To check that it will work, I have added a console.log, console.error and console.info.  Apart from that, it returned a response with a status of 200.

With this I couldn't see any log in Vercel. However, if I return a response with a status other than 200 (I have only tried with 400 and 500), It seems that the error does reach Vercel and along with it, all the previous logs.

Therefore, the current behaviour is that until a response with an error code arrives, the logs that would have been in this execution do not arrive.

Can anyone confirm if this is a bug or an expected behaviour?

**Comment by robpetty** [comment]
Anyone got more insight into handling of logging?  I been debugging what I thought was a race condition for quite a while, but then realized I probably just messed up some supabase queries.  Before I actually make any change to it to fix the issue, I wanted to prove it via data and added some additional debugging to my function.

Now, after finding this post I tried to --follow the logs, but it does not retrieve all the logs.  Many are missing.  When I go back into Vercel and check its logs, I can see a lot more but even there, many are missing (locally in debug mode I will see them for example).

My app is Next.js and using all the regular deployment functionality in Vercel, nothing special.  

To show, here is the --follow:
<img width="621" alt="image" src="https://github.com/vercel/community/assets/1473437/c89f569a-89e8-48a6-968e-1a87a4ee7995">

And here is same section in Vercel (don't let the timestamps fool you, the --follow I believe is in UTC/GMT and web site is in 24 hour/local timezone:

<img width="1276" alt="image" src="https://github.com/vercel/community/assets/1473437/033b9b68-5101-4254-8bde-483a76ec09c4">

One issue that stands out is using console.warn().  I use it a lot for debugging but it seems there is a disconnect between the website, the CLI and node.js with warn/warning.  

- `console.warning()` does not exist for logging, just `console.warn()` (maybe obvious but just pointing out to avoid confusion).
- Per the docs, while not obvious unless you read them (which I admit didn't until I got into this mess), for edge logging you can only log info and error, NOT warn. However, thats not totally true as I can see the warn logs in output (with status XXX). I suspect this is just because the back end is expecting "warn" to be a request response and not a log message.
- But warn logging is not  pulled down by the "`vercel logs {url|id} --follow`"
- However, for requests that fail you can use warning, which is why its available as a choice, BUT....

The filter is also slightly broken at least with auto-complete as warning does not come up as an option though is there as a text hint:

<img width="359" alt="image" src="https://github.com/vercel/community/assets/1473437/c2515891-4a15-4399-a483-c4273db0f267">

<img width="330" alt="image" src="https://github.com/vercel/community/assets/1473437/6f3b5fae-c53a-44d4-b939-ea59d537e209">

So based on the above, I will just switch out to console.error which isn't ideal, to see if that helps my debugging here in general.  Figured I would just ask if anyone had more insight but also share this as it wasn't obvious if your just in the web app and see "warning" and just rolling with it like I was by mistake.

Update:
Meant to add this link:
https://vercel.com/docs/observability/runtime-logs#level

**Comment by dkudrin** [comment]
No logs, ever, on every deployment.

**Reply by j-mcgregor** [reply]
we're experiencing this - do you know if Vercel is now hiding the logs?

**Comment by gigamesh** [comment]
> Therefore, the current behaviour is that until a response with an error code arrives, the logs that would have been in this execution do not arrive.

This seems to be correct.

In my case, the function times out and I get a 504 without any of the error logs because the 504 is generated by Vercel. So its impossible to use logs to debug. 

It's amazing that we have to figure this stuff out through trial and error rather than Vercel communicating it. Moving my API to Firebase.

**Comment by lynixiox** [comment]
For me my computers clock was out of sync with the server
